### DISCUSSION TITLE
Apple to deprecate OpenGL and OpenCL in MacOS 10.14

### DISCUSSION THREAD

**Discussion by Ageless93** [BOINC/boinc#4962] [General]
[Apple deprecates OpenGL and OpenCL support](https://www.phoronix.com/scan.php?page=news_item&px=Apple-Deprecates-OpenGL-OpenCL)

> With today's announcement of macOS 10.14 Mojave, Apple quietly confirmed they are deprecating OpenGL and OpenCL within macOS.

> Apple deprecating OpenCL and OpenGL hardly comes as a surprise given in the past few years they have been pushing their Metal API for graphics and compute across macOS and iOS. Additionally, their OpenGL stack hasn't been updated well in years and has lagged behind the OpenGL 4.x upstream advancements out of The Khronos Group.

> Apps built using OpenGL and OpenCL will continue to run in macOS 10.14, but these legacy technologies are deprecated in macOS 10.14. Games and graphics-intensive apps that use OpenGL should now adopt Metal. Similarly, apps that use OpenCL for computational tasks should now adopt Metal and Metal Performance Shaders.

This means that BOINC needs code enabling it to detect GPUs using Metal.

**Comment by CharlieFenton** [comment]
It usually takes years after an API is deprecated before a new version of mac OS X no longer supports them. Switching from OpenCL and OpenGL to Apple's proprietary Metal APIs would involve far more than the code to detect GPUs. It would mean that all screensavers / project graphics applications would need to be rewritten (perhaps completely) just for Macs, and all project work applications that use GPU computing would need to be rewritten (perhaps completely) just for Macs.

Actually, most or all GLUT APIs were deprecated back in OS 10.9 (October 2013). These are used extensively in the BOINC graphics library: graphics2_unix.cpp, gutil_text.cpp and gutil.cpp.

**Comment by SETIguy** [comment]
I think this will eventually result in many projects not supporting GPU
apps under MacOS.  It also might reduce support for OpenCL in the science
world in favor of CUDA.

On Wed, Jun 6, 2018 at 12:48 AM, Jord van der Elst <notifications@github.com
> wrote:

> Apple deprecates OpenGL and OpenCL support
> <https://www.phoronix.com/scan.php?page=news_item&px=Apple-Deprecates-OpenGL-OpenCL>
>
> With today's announcement of macOS 10.14 Mojave, Apple quietly confirmed
> they are deprecating OpenGL and OpenCL within macOS.
>
> Apple deprecating OpenCL and OpenGL hardly comes as a surprise given in
> the past few years they have been pushing their Metal API for graphics and
> compute across macOS and iOS. Additionally, their OpenGL stack hasn't been
> updated well in years and has lagged behind the OpenGL 4.x upstream
> advancements out of The Khronos Group.
>
> Apps built using OpenGL and OpenCL will continue to run in macOS 10.14,
> but these legacy technologies are deprecated in macOS 10.14. Games and
> graphics-intensive apps that use OpenGL should now adopt Metal. Similarly,
> apps that use OpenCL for computational tasks should now adopt Metal and
> Metal Performance Shaders.
>
> This means that BOINC needs code enabling it to detect GPUs using Metal.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BOINC/boinc/issues/2550>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKXcshA24Y_Un7IcMG7dwYnJw4yG4L2Lks5t54lfgaJpZM4UcJzD>
> .
>



-- 
Eric Korpela
korpela@ssl.berkeley.edu
AST:7731^29u18e3

**Comment by numbermaniac** [comment]
It doesn't help that Metal _still_ doesn't have support for 64-bit integers or floats. If your OpenCL program relies on those, good luck getting it to work on Metal!

https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf

**Comment by CharlieFenton** [comment]
I suspect (hope?) that there will be third-party OpenCL and OpenGL libraries for the Mac to fill this need. FreeGlut is already available, but I understand there are some issues when using it on the Mac. 

CUDA won't help as long as Apple uses AMD GPUs in its Macs.

**Comment by Ferroin** [comment]
MoltenVK (Vulkan on Metal from Khronos) seems relevant: https://github.com/KhronosGroup/MoltenVK

There's also a paid OpenGL on Metal version from the same set of developers who wrote MoltenVK called MoltenGL.

**Comment by AenBleidd** [comment]
@CharlieFenton, any news regarding this? Should we prepare for OpenGL/OpenCL removal from OSX somehow?

**Comment by CharlieFenton** [comment]
It is till supported on MacOS 12.3 on both Intel and Apple Silicon (arm64) Macs. That's all I know. The fact that it's deprecated does not necessarily mean that they will stop including the GLUT and OpenGL libraries. I don't know of any way to learn their intentions. My hope is still that, if and when they do eliminate the libraries, we can find third party libraries and install them with BOINC in a way that projects won't need to make any changes to their worker and graphics executables.

**Comment by AenBleidd** [comment]
Since there is nothing that should be done right now but this topic should be open to track any new updates from Apple, so I decided to convert this from 'Issue' to 'Conversation'.
When OpenGL and/or OpenCL will be deprecated and removed from the OSX default installation, we can think about best approach how to fix that and create a ticket with exact action points.
